### ISSUE TITLE
[BUG] Infinite loop + memory leak in message parsing

### ISSUE THREAD

**Issue by sfPlayer1** [ErdbeerbaerLP/Discord-Chat-Integration#145]
There is a nasty memory leak in `dcshadow.me.vankka.simpleast.core.parser.Parser.parse` (dcintegration 2.0.1) when receiving a complex message similar to `> some non-trivial quoted message (with various tokens) to break the bot.` from Discord. I am not sure whether this exact message triggers it, will investigate further if the obvious approach below fails.

It appears that the parse method enters an infinite loop, gradually filling the heap with presumably queue items or output data at ~100 MB/s.

SimpleAST recently received a couple commits around the relevant code that may fix this and https://github.com/ErdbeerbaerLP/Discord-Chat-Integration/commit/949168da032c4fc9adb4cdff9be2e9ad3861af24 apparently also includes them. Please consider a new dcintegration release with this commit included.

**Comment by ErdbeerbaerLP** [comment]
I already had an different fix implemented in-dev, but not released yet. Will implement that SimpleAST update as soon as I **can**


Link to latest compiled file with workaround implemented: https://cdn.discordapp.com/attachments/302144329785278464/774672955488731136/dcintegration-2.0.2-1.16.jar

btw the problem of that were the newlines

**Comment by ErdbeerbaerLP** [comment]
I just tried using the latest commit of SimpleAST in-dev and removing my workaround. It is still not fixed. But I will go and push an release with the workaround soon

**Comment by sfPlayer1** [comment]
There are a couple non-daemon threads (again), which would at least have a chance to disturb post-crash exit, server thread is long gone in the following dump:

```
"JDA RateLimit-Worker 1" #89 prio=5
idle TIMED_WAITING waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@287c3353
        at java.base@14.0.2/jdk.internal.misc.Unsafe.park(Native Method)
        at java.base@14.0.2/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:252)
        at java.base@14.0.2/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:1661)
        at java.base@14.0.2/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1182)
        at java.base@14.0.2/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:899)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1056)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1116)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base@14.0.2/java.lang.Thread.run(Thread.java:832)

"JDA RateLimit-Worker 2" #90 prio=5
"JDA RateLimit-Worker 3" #96 prio=5
"JDA RateLimit-Worker 4" #114 prio=5
"JDA RateLimit-Worker 5" #142 prio=5
idle WAITING waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@287c3353
        at java.base@14.0.2/jdk.internal.misc.Unsafe.park(Native Method)
        at java.base@14.0.2/java.util.concurrent.locks.LockSupport.park(LockSupport.java:341)
        at java.base@14.0.2/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(AbstractQueuedSynchronizer.java:505)
        at java.base@14.0.2/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3137)
        at java.base@14.0.2/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:1614)
        at java.base@14.0.2/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1177)
        at java.base@14.0.2/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:899)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1056)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1116)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base@14.0.2/java.lang.Thread.run(Thread.java:832)

"Webhook-RateLimit Thread WebhookID: 123" #143 prio=5
idle WAITING waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@25b6f97b
        at java.base@14.0.2/jdk.internal.misc.Unsafe.park(Native Method)
        at java.base@14.0.2/java.util.concurrent.locks.LockSupport.park(LockSupport.java:341)
        at java.base@14.0.2/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(AbstractQueuedSynchronizer.java:505)
        at java.base@14.0.2/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3137)
        at java.base@14.0.2/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:1614)
        at java.base@14.0.2/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1170)
        at java.base@14.0.2/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:899)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1056)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1116)
        at java.base@14.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base@14.0.2/java.lang.Thread.run(Thread.java:832)
```

**Comment by ErdbeerbaerLP** [comment]
This _should_ be fixed by an workaround in version 2.0.2 on CurseForge now
Keeping this open as it only is an workaround
